### PR TITLE
[ewasm] Fix EwasmBuiltinInterpreter to follow the specs

### PR DESCRIPTION
**Pull request by axic**
The specs from here: https://github.com/ewasm/design/blob/master/eth_interface.md

Motivated by this discussion: https://github.com/ethereum/solidity/pull/9508/files#r510443808

Note that the updated tests look "broken", but they are going to be fixed by #9508.

Will squash prior to merge.

### CODE REVIEW

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.h [hunk 1]
```diff
@@ -99,6 +99,9 @@ class EwasmBuiltinInterpreter
 	/// @returns the memory contents (4 bytes) at the provided address (little-endian).
 	/// Does not adjust msize, use @a accessMemory for that
 	uint32_t readMemoryHalfWord(uint64_t _offset);
+	/// @returns the memory contents at the provided address.
```
Comment (aarlt) [inline test/tools/yulInterpreter/EwasmBuiltinInterpreter.h]: the comment seem not to be correct

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp [hunk 2]
```diff
@@ -335,38 +335,38 @@ u256 EwasmBuiltinInterpreter::evalEthBuiltin(string const& _fun, vector<uint64_t
...
 	else if (_fun == "storageStore")
 	{
-		m_state.storage[h256(readU256(arg[0]))] = readU256((arg[1]));
+		m_state.storage[readBytes32(arg[0])] = readBytes32(arg[1]);
 		return 0;
 	}
 	else if (_fun == "storageLoad")
 	{
-		writeU256(arg[1], m_state.storage[h256(readU256(arg[0]))]);
+		writeBytes32(arg[1], m_state.storage[readBytes32(arg[0])]);
```
Comment (aarlt) [inline test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp:365]: maybe we should also call `accessMemory` here and in `storageStore` ?

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp [hunk 3]
```diff
@@ -335,38 +335,38 @@ u256 EwasmBuiltinInterpreter::evalEthBuiltin(string const& _fun, vector<uint64_t
...
 		return 0;
 	}
 	else if (_fun == "getCaller")
 	{
-		// TODO should this only write 20 bytes?
 		writeAddress(arg[0], m_state.caller);
```
Comment (aarlt) [inline test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp:370]: `accessMemory`?

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp (hunk 2 above)
Comment (axic) [reply test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp:365]: `writeBytes32` does that already

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp (hunk 3 above)
Comment (axic) [reply test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp:370]: `writeAddress` does that already

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp (hunk 2 above)
Comment (aarlt) [reply test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp:365]: ah great :)

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.h [hunk 4]
```diff
@@ -109,14 +113,18 @@ class EwasmBuiltinInterpreter
...
 	void writeMemoryByte(uint64_t _offset, uint8_t _value);
 
-	/// Helper for eth.* builtins. Writes to memory (big-endian) and always returns zero.
+	/// Helper for eth.* builtins. Writes to memory (little-endian) and always returns zero.
 	void writeU256(uint64_t _offset, u256 _value, size_t _croppedTo = 32);
 	void writeU128(uint64_t _offset, u256 _value) { writeU256(_offset, std::move(_value), 16); }
-	void writeAddress(uint64_t _offset, u256 _value) { writeU256(_offset, std::move(_value), 20); }
-	/// Helper for eth.* builtins. Reads from memory (big-endian) and returns the value;
+	/// Helper for eth.* builtins. Writes to memory (as a byte string).
+	void writeBytes32(uint64_t _offset, util::h256 _value) { accessMemory(_offset, 32); writeMemory(_offset, _value.asBytes()); }
```
Comment (leonardoalt) [inline test/tools/yulInterpreter/EwasmBuiltinInterpreter.h:122]: Why have both `writeBytes32` and `writeU256`?

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp [hunk 5]
```diff
@@ -545,17 +561,17 @@ void EwasmBuiltinInterpreter::writeU256(uint64_t _offset, u256 _value, size_t _c
 	accessMemory(_offset, _croppedTo);
 	for (size_t i = 0; i < _croppedTo; i++)
 	{
-		m_state.memory[_offset + _croppedTo - 1 - i] = uint8_t(_value & 0xff);
+		m_state.memory[_offset + i] = uint8_t(_value & 0xff);
```
Comment (leonardoalt) [inline test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp:564]: Why was `_croppedTo - 1` not necessary?

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.h (hunk 4 above)
Comment (axic) [reply test/tools/yulInterpreter/EwasmBuiltinInterpreter.h:122]: Different byte ordering.

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp (hunk 5 above)
Comment (axic) [reply test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp:564]: It did an endianness swap, wrongly.

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.h (hunk 4 above)
Comment (leonardoalt) [reply test/tools/yulInterpreter/EwasmBuiltinInterpreter.h:122]: Could that be made more explicit in the function's names? Maybe in the little-endian case only

Comment (axic) [reply test/tools/yulInterpreter/EwasmBuiltinInterpreter.h:122]: These names follow the "ewasm spec".

File: test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp [hunk 6]
```diff
@@ -523,6 +523,12 @@ uint32_t EwasmBuiltinInterpreter::readMemoryHalfWord(uint64_t _offset)
 	return r;
 }
 
+void EwasmBuiltinInterpreter::writeMemory(uint64_t _offset, bytes _value)
```
Comment (chriseth) [inline test/tools/yulInterpreter/EwasmBuiltinInterpreter.cpp]: `const&`

**Review by chriseth — DISMISSED** [review]
Could add a `const&` but can also merge as is.

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
> Could add a const& but can also merge as is.

Done.

**Pull request merged**